src/SearchField/index.jsx

```diff
@@ -0,0 +1,278 @@
+      <Button
+        className={[classNames(
+          styles['border-0'],
+          styles['ml-1'],
+          styles['search-field-clear-btn'],
+        )]}
+        label={(
+          <small>
+            <Icon
+              className={[classNames(
+                FontAwesomeStyles.fa,
+                FontAwesomeStyles['fa-times'],
+                {
+                  [styles['text-light']]: !focused,
+                  [styles['text-muted']]: focused,
+                },
+              )]}
+              id={newId('icon-SearchField')}
+              screenReaderText="clear search"
+            />
+          </small>
+        )}
+        onClick={this.handleClear}
+        aria-hidden={!shouldShowClearButton}
+      />
+    );
+  }
+
+  getSearchButton() {
+    const { focused } = this.state;
+    const {
+      hasRoundedCorners,
+      hasClearButton,
+      searchButtonLabel,
+    } = this.props;
+    const inputTextHasValue = this.inputTextHasValue();
+
+    return (
+      <Button
+        className={[classNames(
+          styles['search-field-search-btn-transition'],
+          {
+            [styles['ml-1']]: !hasClearButton,
+            [styles['text-muted']]: !focused && inputTextHasValue,
+            [styles['border-left']]: !focused && inputTextHasValue,
+            [styles['btn-outline-primary']]: focused && inputTextHasValue,
+            [styles['rounded-0']]: !hasRoundedCorners,
+            [styles['search-field-rounded']]: hasRoundedCorners,
+            [styles['search-field-rounded-top-left-0']]: hasRoundedCorners,
+            [styles['search-field-rounded-bottom-left-0']]: hasRoundedCorners,
+          },
+        )]}
+        label={searchButtonLabel}
+        disabled={!this.inputTextHasValue()}
+        inputRef={(input) => { this.searchButton = input; }}
+        onClick={this.handleSubmit}
+      />
+    );
+  }
+
+  getSearchActionButtons() {
+    const buttons = [
+      this.getSearchButton(),
+    ];
+
+    if (this.shouldShowClearButton()) {
+      buttons.unshift(this.getClearButton());
+    }
+
+    return buttons;
+  }
+
+  shouldShowClearButton() {
+    return this.props.hasClearButton && this.inputTextHasValue();
+  }
+
+  inputTextHasValue() {
+    const { value } = this.state;
+    return value && value.trim().length > 0;
+  }
+
+  handleFocus(event) {
+    this.setState({ focused: true });
+    this.props.onFocus(event);
+  }
+
+  handleBlur(event) {
+    this.setState({ focused: false });
+    this.props.onBlur(event);
+  }
+
+  handleChange(value) {
+    this.setState({ value });
+    this.props.onChange(value);
+  }
+
+  handleClear() {
+    this.setState({ value: '' });
+    this.props.onChange('');
+    this.textInput.focus();
+  }
+
+  handleKeyPress(event) {
+    if (event.key === 'Enter' && this.inputTextHasValue()) {
+      this.handleSubmit();
+    }
+  }
+
+  handleSubmit() {
+    const { value } = this.state;
+    this.searchButton.focus();
+    this.props.onSubmit(value.trim());
```

Comment:
Why `trim`? An input value that contains whitespaces is still a valid input, right? Shouldn't it be left to the caller of this API how they deal with whitespaces?

Reply:
What value would leaving leading/trailing whitespaces here have for the end user? I can't think of a case where the caller of this API would actually want whitespaces, in which case, doesn't it make sense for `handleSubmit` to remove that step to clean the input?

Reply:
There are two reasons why

1. It can lead to inconsistencies with the user experience. Imagine this is a username input field and imagine that I can actually create usernames with whitespaces. I type in "username ". The username that's actually created is "username" because this component stripped out the whitespace and the API to create usernames used the value from the `onSubmit` prop. Here's another one - what if I want to display a max character length warning for some input value?
2. It's arbitrarily opinionated. What if I can't think of a business case (right now) for having the letter `a` in my input? Should I then strip out the letter `a` from my input? Doing formatting like this unnecessarily boxes in the API user - let the API user decide how they'd like to format the input value.